http_client: implement NO_PROXY support

erain opened the pull request:
implement NO_PROXY support in fluent-bit:
- `NO_PROXY` is a standard environment variable for the URL should not go through the proxy, when `HTTP_PROXY` is used. 
- `NO_PROXY` is "A comma-separated list of host names that shouldn't go through any proxy is set in (only an asterisk, '*' matches all hosts)" ([reference](https://superuser.com/questions/944958/are-http-proxy-https-proxy-and-no-proxy-environment-variables-standard))

FIXES https://github.com/fluent/fluent-bit/issues/2805

<!-- Provide summary of changes -->

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change
<!-- Invoke Fluent Bit and Valgrind as: $ valgrind ./bin/fluent-bit <args> -->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [ ] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

erain commented:
I successfully tested this feature in our environment, @pranavmarla could you help also test in your environment?

My testing pods are running in a kubernetes environment with environment variables something like below, with a squid http proxy. And please note there should not be spaces in the `NO_PROXY` string.

```
        - name: HTTP_PROXY
          value: http://abc:123@<proxy-host>:<proxy-port>
        - name: NO_PROXY
          value: 127.0.0.1,localhost,kubernetes.default.svc.cluster.local
```

qingling128 commented:
Hi @edsiper , could you help prioritize the code review of this PR? It is required for our GA as well.

pranavmarla commented:
> I successfully tested this feature in our environment, @pranavmarla could you help also test in your environment?
> 
> My testing pods are running in a kubernetes environment with environment variables something like below, with a squid http proxy. And please note there should not be spaces in the `NO_PROXY` string.
> 
> ```
>         - name: HTTP_PROXY
>           value: http://abc:123@<proxy-host>:<proxy-port>
>         - name: NO_PROXY
>           value: 127.0.0.1,localhost,kubernetes.default.svc.cluster.local
> ```

@erain That's awesome, glad to hear it!

Unfortunately though, it might be some time till I'm in a position to test this out -- my EKS environment, Fluent Bit setup, etc. has long since been torn down, so I'd need to recreate all that.
Also, to test out your changes, I assume I'd need to build Fluent Bit from source (i.e. your branch) and then create a custom Docker image containing that custom Fluent Bit to reference in my Kubernetes deployment, all of which will also take time.
So basically, please don't wait on me to merge this change!

edsiper commented on src/flb_upstream.c:
```diff
@@ -237,6 +237,39 @@ struct flb_upstream *flb_upstream_create(struct flb_config *config,
+/*
+ * Checks whehter a destinate URL should be proxied.
+ */
+int flb_should_proxy_for_host(const char *host, const char *proxy, const char *no_proxy)
```
the function name does not follow the interface pattern, please rename it to something like ```flb_upstream_needs_proxy()```

erain replied:
DONE.

edsiper commented on src/flb_upstream.c:
```diff
@@ -237,6 +237,39 @@ struct flb_upstream *flb_upstream_create(struct flb_config *config,
+    }
+
+    /* check the URL list in the NO_PROXY  */
+    char *no_proxy_url = strtok(no_proxy, ",");
```
- variable needs to be declared on top

- do not use strtok, instead go with flb_slist_split_tokens() API

erain replied:
DONE. Thanks for letting me know, I used `flb_slist_split_string` for this now and also enhanced its unit tests.

edsiper commented on include/fluent-bit/flb_str.h:
```diff
@@ -58,4 +58,13 @@ static inline char *flb_strndup(const char *s, size_t n)
 }
 
+/* emptyval checks whether a string has a non-null value "". */
+static inline int flb_str_emptyval(const char *s)
```
this function is only called from flb_config.c file, let's move it there in the .c file on top as a static inline.

erain replied:
DONE.

erain commented:
Hi @edsiper , I've pushed a new commit which addressed all comments, and I also tested my new commit the same way as https://github.com/fluent/fluent-bit/pull/3272#issuecomment-805153418

frcai commented on include/fluent-bit/flb_config.h:
```diff
@@ -154,6 +154,16 @@ struct flb_config {
+     * Example: NO_PROXY="127.0.0.1,localhost,kubernetes.default.svc"
+     * Note: only `,` is allowed as seperator between URLs.
+     */;
+    char *no_proxy;
```
Can we support IP ranges in CIDR format? Check https://docs.openshift.com/enterprise/3.2/install_config/http_proxies.html?

erain replied:
I don't think this is feasible for now. This is because that we only have `hostname` when creating a `http_upstream`. I don't know a easy / feasible way for turning that into an IP address and do the CIDR check. 

(Golang supports that natively. Not so easy with the fluent-bit's C code base. )

If you have some ideas, feel free to enhance this feature based on my PR.

frcai replied:
I think if IP address was configured in config file, the value of hostname should be IP address which you check against with CIDR.

erain replied:
The hostname is something uncertain, depends on the remote host. For example, for stackdriver output plugin it will be the `logging.googleapis.com`. Almost certainly we will have a domain name instead of a IP address here, and since the whole fluent-bit's http client is implemented in its own codebase, doing DNS lookup etc. here would be non-trivial. 

So again, though this is nice to have, I would consider this as a separate feature requirement. Feel free to contribute PR for this.

edsiper commented:
hi @erain, after code review and CI fixes, is this ready to be merged from a functionality perspective ?

erain commented:
> hi @erain, after code review and CI fixes, is this ready to be merged from a functionality perspective ?

Yes it is. 

Sorry that I was occupied in the past few days and haven't looked into the CI failed earlier. It looks like related to my attempt on enhancing the current unit test, and I believe I followed the existing pattern. Could you help me to see what did I miss?

edsiper commented:
- This one is missing the signature on commits (DCO error) 
- Looks like there is a header broken ? check this: https://ci.appveyor.com/project/fluent/fluent-bit-2e87g/builds/38503081/job/aa7rivpn34uor77w#L1380
- unit test are broken: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2298240264#step:5:1431

erain commented:
Thanks @edsiper , I will do another commit ASAP (currently busy with other on-fire issues on my side)

JeffLuoo commented on include/fluent-bit/flb_config.h:
```diff
@@ -154,6 +154,16 @@ struct flb_config {
+     * and respected by `NO_PROXY` environment variable when `HTTP_PROXY` is used.
+     * Example: NO_PROXY="127.0.0.1,localhost,kubernetes.default.svc"
+     * Note: only `,` is allowed as seperator between URLs.
+     */;
```
Hi Yu @erain Is this `;` a typo?

erain replied:
FIxed. PTAL.

JeffLuoo commented:
Hi Yu @erain Could you please remove the `;` as https://github.com/fluent/fluent-bit/pull/3272#discussion_r620826008 and re-commit to trigger the test again? 

Also don't forget the signoff `-s` option when committing : D (details: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2241296333)

erain commented:
> Hi Yu @erain Could you please remove the `;` as [#3272 (comment)](https://github.com/fluent/fluent-bit/pull/3272#discussion_r620826008) and re-commit to trigger the test again?
> 
> Also don't forget the signoff `-s` option when commit : D (details: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2241296333)

Many thanks!!

I fixed changes and now CI should pass. And I will prepare a doc PR

JeffLuoo commented:
Ping @edsiper  Could you please take a look at this PR? Thanks

qingling128 commented:
@agup006 @edsiper Any idea about the failed integration test https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2462035748

qingling128 commented:
Re-triggered the failed test. Running now.

qingling128 commented:
`Run integration tests for PR / publish the docker images for PR (pull_request_target)` seems to have failed again: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2521701599. It successfully built the container, but failed to push it to Dockerhub. That seems like a test infra issue though, probably not triggered by this PR.

qingling128 commented:
You are welcome! Do we have any planned release coming that would include this feature?

edsiper commented:
we can do something early next week.

qingling128 commented:
Sounds great. Thanks @edsiper @erain @JeffLuoo for making this happen!
